### PR TITLE
(PUP-10142) Allow Puppet settings redefinition

### PR DESCRIPTION
**Pull request by tvpartytonight**

### CODE REVIEW

File: lib/puppet.rb [hunk 1]
```diff
@@ -60,6 +60,14 @@ def self.settings
     @@settings
   end
 
+  # The puppetserver project has its own settings class that is thread-aware; this
+  # method is here to allow the puppetserver to define its own custom settings class
+  # for multithreaded puppet. It is not intended for use outside of the puppetserver
+  # implmentation.
+  def self.replace_settings_object(new_settings)
```
Comment (joshcooper) [inline lib/puppet.rb]: maybe `def self.settings=(new_settings)` to match the reader just above? Could you add a test to `spec/unit/settings_spec.rb`?

**Comment by mwaggett** [comment]
This will need a rebase to fix that Ruby 2.6 test failure.

File: lib/puppet.rb (hunk 1 above)
Comment (tvpartytonight) [reply lib/puppet.rb]: Sure, I can add the spec test.

I named it this funny name to indicate this is something we don't expect people to really use; but I can change it to the more normal name if that makes better sense. It also felt like an easy typo to make, `Puppet.settings=` vs `Puppet.settings[:yourcoolsetting]=`.

Comment (Magisus) [reply lib/puppet.rb]: I think I'm with Tony on this one for the name. `settings=` seems likely to send the wrong message on how to interact with this class.

Comment (justinstoller) [reply lib/puppet.rb]: This is similar convo to what we had regarding `global_context_set_unsafe_whatever_we_ended_up_on` and I'd agree with Tony and Maggie that a normal setter looks too much like something open for use. Maybe throwing an `@api private` on here would help?

But I agree with Cooper it should have a test regardless of its name.

Comment (tvpartytonight) [reply lib/puppet.rb]: Name unchanged, test added.

**Pull request merged**